### PR TITLE
Add new Pipelined Kernels sample

### PR DESCRIPTION
**Pull request by intel-jisheng1**
# Adding a New Sample(s)

## Description

Please include a description of the sample

## Checklist
Administrative
- [ ] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): <insert Name Here>
- [ ] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager

Code Development
- [X] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [X] Adhere to readme template 
- [X] Enforce format via clang-format config file
- [X] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [X] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [X] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

**Comment by jarrodblackburn** [comment]
I recommend adding a block diagram that goes with the timing diagram showing the port directions since I was having a hard time figuring out the ready port pairing with the other ports.  I'm guessing start, ready_out, and arg_input are paired together to push arguments into the kernel and done and ready_in are paired together to notify the controller of the kernel completions.  If those groupings are correct I recommend grouping them similarly in the timing diagram as well.

In the kernel .tcl file what does this interface map to?  The start/done bits behave like 'valid' and the kernel arguments are just streaming data so this can show up in platform designer as a pair of simple Avalon streams (and AXI streams at a later date).  If these are still crammed into a conduit we should probably talk since that's not a user friendly way to expose an interface given it was made up for oneAPI and so there is nothing else that natively supports it unlike if you use Avalon/AXI-ST to accomplish the same thing.

**Comment by intel-jisheng1** [comment]
> In the kernel .tcl file what does this interface map to? The start/done bits behave like 'valid' and the kernel arguments are just streaming data so this can show up in platform designer as a pair of simple Avalon streams (and AXI streams at a later date). If these are still crammed into a conduit we should probably talk since that's not a user friendly way to expose an interface given it was made up for oneAPI and so there is nothing else that natively supports it unlike if you use Avalon/AXI-ST to accomplish the same thing.

Here are the relevant sections in the tcl file:

```
####  ZTS4MyIP_arg_input
add_interface ZTS4MyIP_arg_input conduit end
set_interface_property ZTS4MyIP_arg_input ENABLED true
set_interface_property ZTS4MyIP_arg_input associatedClock clock
set_interface_property ZTS4MyIP_arg_input associatedReset resetn
add_interface_port ZTS4MyIP_arg_input ZTS4MyIP_arg_input data input 64

#### The component invocation protocol interface (start)
add_interface ZTS4MyIP_streaming_start conduit sink
set_interface_property ZTS4MyIP_streaming_start ENABLED true
set_interface_property ZTS4MyIP_streaming_start associatedClock clock
set_interface_property ZTS4MyIP_streaming_start associatedReset resetn
add_interface_port ZTS4MyIP_streaming_start ZTS4MyIP_streaming_start valid input 1
add_interface_port ZTS4MyIP_streaming_start ZTS4MyIP_streaming_ready_out ready output 1

#### The component return protocol interface (done)
add_interface ZTS4MyIP_streaming_done conduit source
set_interface_property ZTS4MyIP_streaming_done ENABLED true
set_interface_property ZTS4MyIP_streaming_done associatedClock clock
set_interface_property ZTS4MyIP_streaming_done associatedReset resetn
add_interface_port ZTS4MyIP_streaming_done ZTS4MyIP_streaming_done valid output 1
add_interface_port ZTS4MyIP_streaming_done ZTS4MyIP_streaming_ready_in ready input 1
```

Is this what you are expecting? @jarrodblackburn

**Comment by jarrodblackburn** [comment]
Those are conduit interfaces but they are tagged with signal roles so I'm not sure what Platform Designer will do with that (never seen that before).  I suspect those roles of 'data' 'ready' and 'valid' only work within the conduit and will not be wired up to Avalon-ST or AXI-ST streaming interfaces without a custom gasket in between when using system building tools like Platform Designer.

If it was designed this way because we don't know how many arguments a kernel will have this could have been solved using Avalon-ST and 'data', 'valid', 'ready', and 'eop'.  The arguments could have been streamed to the kernel using 'data' and the start condition signaled with 'eop'.  Then if there is only one argument then you just provide the data and assert 'eop' at the same time to send the argument and start condition.  Something to think about for enhancing this interface since it seems to me to fit into Avalon-ST or AXI-ST very well in my opinion.  That said, since this is using conduits all the more reason to include a block diagram with this example since it'll be a new interface for most users.

**Comment by bsyrowik** [comment]
I believe the _hw.tcl should read as follows to more closely match the intent here:

```
####  ZTS4MyIP_arg_input
add_interface ZTS4MyIP_arg_input conduit end
...

#### The component invocation protocol interface (start)
add_interface ZTS4MyIP_streaming_start avalon_streaming sink
...

#### The component return protocol interface (done)
add_interface ZTS4MyIP_streaming_done avalon_streaming source
...
```

**Comment by bsyrowik** [comment]
I believe the expectation here is that any component arguments marked as "conduit" in the .cpp source will show up as separate conduit interfaces in the _hw.tcl (which is what we see).  It is up to the user to ensure they are properly synchronized with the *_streaming_start interface (which is essentially a data-less avalon streaming sink).  I _think_ this is how it worked in HLS as well?  (Note: I think our _hw.tcl is currently wrong for the start and done interfaces, but from what I've seen platform designer seems to understand the interface as AVST anyway?  Although, this is only from my experience working with the sim flow, so my understanding may be incorrect...).

If a user wants an AVST interface for their inputs they can use a host pipe, and if they want an AVMM interface they can use a mmhost interface type in their .cpp file.

Jarrod: does this make sense from a user's perspective, or is what we're creating a bit cumbersome/unintuitive (or plain wrong!)?

**Comment by jarrodblackburn** [comment]
I think as long as this is well documented and the user can swap it out for other interface types like AvMM then customers can make the interface decision for themselves.  I think having the kernel arguments sent over a conduit but the kernel invocations using AvST would be a bit strange given you normally don't want multiple paths potentially haven't different latencies in the control plane that need to be synchronized.  So if the user can request that the kernel arguments be sent over the same stream as the kernel start condition then I have no concerns that users can optionally break the arguments out to a separate interfaces such as conduits.  

If keeping the kernel arguments and start condition in the same stream is not possible today then perhaps that can be added in the future.  If a kernel can have so many arguments that it won't fit into a single streaming beat then the start condition can be qualified by EOP to ensure the arguments are kept synchronized inband.  Likewise if we end up supporting return types the same can be applied to the done stream where the return value flows inband with the done.  What I described would also port 1:1 to AXI-ST as well since EOP can be mapped to either TLAST or TUSER.

**Comment by bsyrowik** [comment]
> 

Thank you very much for the thorough comments Jarrod!  It sounds like we may need to do a bit of work to make the purely-streaming components more user-friendly.  Bob and I will make sure your advice and concerns are relayed to the appropriate people :)

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/pipelined_kernels/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    # Force CMake to use icpx rather than the default C++ compiler/linker 
+    # (needed on Windows only)
+    include (CMakeForceCompiler)
+    CMAKE_FORCE_CXX_COMPILER (icx-cl IntelDPCPP)
+    include (Platform/Windows-Clang)
+endif()
+
+cmake_minimum_required (VERSION 3.4)
+
+project(register_map_and_streaming_interfaces CXX)
```
Comment (yuguen) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/pipelined_kernels/CMakeLists.txt]:
```suggestion
project(pipelined_kernels CXX)
```

**Review by yuguen — CHANGES_REQUESTED** [review]

**Review by yuguen — APPROVED** [review]

**Pull request closed without merging**